In [1]:
google_colab_mode= False

# Just for Google Colab

In [ ]:
import os
from getpass import getpass
import urllib

user = input('User name: ')
password = getpass('Password: ')
password = urllib.parse.quote(password) # your password is converted into url format

cmd_string = 'git clone https://{0}:{1}@github.com/GianCarloMilanese/dsim_project.git'.format(user, password)

os.system(cmd_string)
cmd_string, password = "", "" # removing the password from the variable
google_colab_mode=True

In [ ]:
!ls -lh dsim_project/

In [ ]:
! git clone https://github.com/Jakobovski/free-spoken-digit-dataset.git && mv free-spoken-digit-dataset/recordings dsim_project/

In [ ]:
!ls -lh dsim_project

In [ ]:
import sys
sys.path.insert(1, "dsim_project/Audio")

# Import libraries

In [ ]:
google_colab_mode=False

In [3]:
import cnn_models
import data_preparation
from sklearn.metrics import classification_report, confusion_matrix
import numpy as np
from sklearn.svm import SVC
import tensorflow as tf
import data_augmentation
import random

/Users/kappa/opt/miniconda3/envs/dsim/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/Users/kappa/opt/miniconda3/envs/dsim/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/Users/kappa/opt/miniconda3/envs/dsim/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/Users/kappa/opt/miniconda3/envs/dsim

# Fix seed

In [4]:
SEED = 10
random.seed(SEED)
tf.random.set_random_seed(SEED)# if working on tf < 2.0
#tf.random.set_seed(SEED)

# Load recordings
## STANDARD RECORDINGS - No spectrogram normalization

In [5]:
if google_colab_mode:
    fsdd_dir="dsim_project/Audio/recordings"
    our_recs_dir="dsim_project/Audio/preprocessed_recs"
else:
    fsdd_dir="./recordings/"
    our_recs_dir="./preprocessed_recs/"

In [6]:
recordings = data_preparation.load_recordings(paths=[fsdd_dir, our_recs_dir])

Loading from ./recordings/



Loading from ./preprocessed_recs/


Raw recordings have different lengths? Let's check it out:

In [7]:
min_y = min(map(np.shape, recordings))[0]
max_y = max(map(np.shape, recordings))[0]
print(min_y, max_y)

1010 18262


The difference is quite huge! Let's see which are the longest recordings:

In [8]:
a = [len(x) for x in recordings]
a.sort(reverse=True)
a[0:10]

[18262, 17567, 9015, 8995, 8435, 8281, 8201, 8068, 7755, 7356]

Two recordings have length 18262 and 17567, while the others are around 20K. Let's identify them:

In [9]:
a = [len(x) for x in recordings]
first_length=18262
second_length=17567
index_first = a.index(first_length)
index_second = a.index(second_length)

In [10]:
len(recordings[index_first])

18262

In [11]:
len(recordings[index_second])

17567

I have found them. For knowing to which digit and speaker they are associated I first need to load the labels:

In [12]:
labels_speakers = data_preparation.load_labels(paths=[fsdd_dir, our_recs_dir], label_type="speakers")
labels_digits = data_preparation.load_labels(paths=[fsdd_dir, our_recs_dir])

In [13]:
print("Longest track is associated with speaker {}, digit {}".format(labels_speakers[index_first],labels_digits[index_first]))
print("Second longest track is associated with speaker {}, digit {}".format(labels_speakers[index_second],labels_digits[index_second]))

Longest track is associated with speaker theo, digit 9
Second longest track is associated with speaker theo, digit 7


So the problem is with theo, which has 500 recordings, digit 9 and 7, which respectively have 200 recordings. We can safely delete them and saving to pad many thousands of 0s (there will be (18262 - 9015) less zeros)

In [14]:
max_track_length=17000 # it will be useful later on
print("Before: {}".format(len(recordings)))
recordings=np.delete(recordings,[index_first, index_second])
print("After: {}".format(len(recordings)))

Before: 2400
After: 2398


In [15]:
print("Before: {}".format(len(labels_speakers)))
labels_speakers=np.delete(labels_speakers,[index_first, index_second])
print("After: {}".format(len(labels_speakers)))

Before: 2400
After: 2398


In [16]:
print("Before: {}".format(len(labels_digits)))
labels_digits=np.delete(labels_digits,[index_first, index_second])
print("After: {}".format(len(labels_digits)))

Before: 2400
After: 2398


Let's now double check to see if everything went well. Now the longest recording will be around 9 K

In [17]:
a = [len(x) for x in recordings]
a.sort(reverse=True)
a[0:10]

[9015, 8995, 8435, 8281, 8201, 8068, 7755, 7356, 7147, 7038]

Yes! However the recordings have all different lengths: for this reason we can add 0s at the beginning and at the end in order to uniform them

In [18]:
pad_recordings = data_preparation.pad_zeros(recordings)

pad_zeros >>>
pad_zeros <<<


What is the range now?

In [19]:
min_y = min(map(np.shape, pad_recordings))[0]
max_y = max(map(np.shape, pad_recordings))[0]
print(min_y, max_y)

9015 9015


We can now compute spectograms:

In [20]:
spects = [data_preparation.compute_spectrogram(x) for x in pad_recordings]
spects = np.array(spects)

Let's also compute "normalized spectrograms

In [21]:
norm_spects = [data_preparation.compute_spectrogram(x, normalize=True) for x in pad_recordings]
norm_spects = np.array(norm_spects)

## Augmentation

In [22]:
%%time
X_train_digit, y_train_digit, X_val_digit, y_val_digit, X_test_digit, y_test_digit = data_preparation.prepare_augmented_recordings(audio_dirs= [fsdd_dir, our_recs_dir],
                             y_type= ['digit', 'digit'],
                             n_category_test=15,
                             include_pitch=True,
                             max_length=max_track_length)

split_and_augment_dataset >>>
enrich_dataset>>>
Max length: 17000, shape:(17567,)
Max length: 17000, shape:(18262,)
enrich_dataset <<<
split_and_augment_dataset <<<
split_and_augment_dataset >>>
enrich_dataset>>>
enrich_dataset <<<
split_and_augment_dataset <<<
conversion_done!
compute_spectrograms >>>
9015
pad_zeros >>>
pad_zeros <<<
pad_zeros >>>
pad_zeros <<<
pad_zeros >>>
pad_zeros <<<
Padding done
compute_spectrograms <<<
CPU times: user 5min 37s, sys: 14 s, total: 5min 51s
Wall time: 4min 59s


In [23]:
print("Lengths : {}, {}, {}, {}, {}, {}".format(len(X_train_digit),
                                                len(y_train_digit),
                                                len(X_val_digit),
                                                len(y_val_digit),
                                                len(X_test_digit),
                                                len(y_test_digit),))

Lengths : 18462, 18462, 4616, 4616, 300, 300


In [24]:
%%time
X_train_speaker, y_train_speaker, X_val_speaker, y_val_speaker, X_test_speaker, y_test_speaker = data_preparation.prepare_augmented_recordings(
    audio_dirs= [our_recs_dir, fsdd_dir],
    y_type= ['speakers_us', 'speakers_default'],
    n_category_test=30,
    include_pitch=False,
    max_length=max_track_length)

split_and_augment_dataset >>>
enrich_dataset>>>
enrich_dataset <<<
split_and_augment_dataset <<<
split_and_augment_dataset >>>
enrich_dataset>>>
Max length: 17000, shape:(17567,)
Max length: 17000, shape:(18262,)
enrich_dataset <<<
split_and_augment_dataset <<<
conversion_done!
compute_spectrograms >>>
9015
pad_zeros >>>
pad_zeros <<<
pad_zeros >>>
pad_zeros <<<
pad_zeros >>>
pad_zeros <<<
Padding done
compute_spectrograms <<<
CPU times: user 4min 9s, sys: 11.7 s, total: 4min 21s
Wall time: 5min 7s


In [25]:
print("Lengths : {}, {}, {}, {}, {}, {}".format(len(X_train_speaker),
                                        len(y_train_speaker),
                                        len(X_val_speaker),
                                        len(y_val_speaker),
                                        len(X_test_speaker),
                                        len(y_test_speaker)))

Lengths : 10358, 10358, 2590, 2590, 240, 240


# Standard recordings
## Numbers

Split data in train, val and test

In [26]:
X_train, X_val, X_test, y_train, y_val, y_test = data_preparation.split_train_test_baseline_spectrograms(spects, labels_digits)

In [27]:
clf1 = SVC(kernel='rbf', class_weight='balanced', gamma="auto")

In [28]:
%%time
clf1 = clf1.fit(X_train, y_train)

CPU times: user 27.7 s, sys: 222 ms, total: 27.9 s
Wall time: 29.7 s


In [29]:
%%time
y_pred = clf1.predict(X_val)
print(classification_report(y_val, y_pred))

              precision    recall  f1-score   support

           0       0.83      0.12      0.20        43
           1       0.41      0.35      0.38        43
           2       0.53      0.17      0.25        48
           3       0.62      0.23      0.34        56
           4       0.23      0.44      0.30        43
           5       0.81      0.35      0.49        48
           6       0.15      0.64      0.24        55
           7       0.88      0.27      0.41        56
           8       0.85      0.23      0.37        47
           9       0.67      0.49      0.56        41

    accuracy                           0.33       480
   macro avg       0.60      0.33      0.35       480
weighted avg       0.60      0.33      0.35       480

CPU times: user 5.59 s, sys: 47.8 ms, total: 5.63 s
Wall time: 6.11 s


### Normalize spectrograms

In [30]:
X_train, X_val, X_test, y_train, y_val, y_test = data_preparation.split_train_test_baseline_spectrograms(norm_spects, labels_digits)

In [31]:
%%time
clf1 = SVC(kernel='rbf', class_weight='balanced', gamma="auto")
clf1 = clf1.fit(X_train, y_train)

CPU times: user 17.2 s, sys: 200 ms, total: 17.4 s
Wall time: 20.2 s


In [32]:
%%time
y_pred = clf1.predict(X_val)
print(classification_report(y_val, y_pred))

              precision    recall  f1-score   support

           0       0.93      0.93      0.93        43
           1       0.82      0.72      0.77        43
           2       0.57      0.92      0.70        48
           3       0.91      0.52      0.66        56
           4       0.94      0.72      0.82        43
           5       0.95      0.77      0.85        48
           6       0.63      0.84      0.72        55
           7       0.80      0.88      0.84        56
           8       0.84      0.66      0.74        47
           9       0.81      0.93      0.86        41

    accuracy                           0.78       480
   macro avg       0.82      0.79      0.79       480
weighted avg       0.82      0.78      0.78       480

CPU times: user 5.3 s, sys: 62.3 ms, total: 5.36 s
Wall time: 5.88 s


### CNNs

#### Normalized spectrograms

In [33]:
X_train, X_val, X_test, y_train, y_val, y_test, input_shape = data_preparation.split_train_test_nn(norm_spects, labels_digits)

In [35]:
model = cnn_models.paper_architecture(10, input_shape=input_shape)

Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 63, 27, 32)        544       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 30, 12, 32)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 14, 5, 64)         32832     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 6, 1, 64)          0         
_________________________________________________________________
flatten (Flatten)            (None, 384)               0         
_________________________________________________________________
dense (Dense)        

In [36]:
callback = tf.keras.callbacks.EarlyStopping(monitor='val_loss', restore_best_weights=True, patience=3)

In [37]:
%%time
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val, y_val))

Train on 1438 samples, validate on 480 samples
Instructions for updating:
Use tf.cast instead.
Epoch 1/10
1438/1438 [==============================] - 3s 2ms/sample - loss: 2.2941 - acc: 0.1189 - val_loss: 2.2513 - val_acc: 0.1500
Epoch 2/10
1438/1438 [==============================] - 3s 2ms/sample - loss: 2.2452 - acc: 0.1516 - val_loss: 2.1974 - val_acc: 0.2146
Epoch 3/10
1438/1438 [==============================] - 3s 2ms/sample - loss: 2.1722 - acc: 0.2316 - val_loss: 2.0930 - val_acc: 0.3313
Epoch 4/10
1438/1438 [==============================] - 3s 2ms/sample - loss: 2.0813 - acc: 0.2656 - val_loss: 1.9382 - val_acc: 0.4458
Epoch 5/10
1438/1438 [==============================] - 3s 2ms/sample - loss: 1.9531 - acc: 0.2928 - val_loss: 1.7715 - val_acc: 0.4646
Epoch 6/10
1438/1438 [==============================] - 3s 2ms/sample - loss: 1.7468 - acc: 0.3679 - val_loss: 1.5091 - val_acc: 0.4896
Epoch 7/10
1438/1438 [==============================] - 3s 2ms/sample - loss: 1.6246 - ac

In [38]:
y_val_nn = np.argmax(y_val, axis=1) # Convert one-hot to index
y_pred = model.predict_classes(X_val)
print(classification_report(y_val_nn, y_pred))

              precision    recall  f1-score   support

           0       0.68      0.74      0.71        43
           1       0.50      0.56      0.53        43
           2       0.43      0.73      0.54        48
           3       0.83      0.09      0.16        56
           4       0.80      0.81      0.80        43
           5       1.00      0.48      0.65        48
           6       0.56      0.67      0.61        55
           7       0.70      0.62      0.66        56
           8       0.52      0.55      0.54        47
           9       0.50      0.78      0.61        41

    accuracy                           0.59       480
   macro avg       0.65      0.60      0.58       480
weighted avg       0.66      0.59      0.57       480



#### Standard spectrogram

In [39]:
X_train, X_val, X_test, y_train, y_val, y_test, input_shape = data_preparation.split_train_test_nn(spects, labels_digits)

In [40]:
model = cnn_models.paper_architecture(10, input_shape)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_2 (Conv2D)            (None, 63, 27, 32)        544       
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 30, 12, 32)        0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 14, 5, 64)         32832     
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 6, 1, 64)          0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 384)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 100)               38500     
_________________________________________________________________
dropout_1 (Dropout)          (None, 100)               0         
__________

In [41]:
%%time
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val, y_val))

Train on 1438 samples, validate on 480 samples
Epoch 1/10
1438/1438 [==============================] - 3s 2ms/sample - loss: 2.5060 - acc: 0.1224 - val_loss: 2.2313 - val_acc: 0.1458
Epoch 2/10
1438/1438 [==============================] - 2s 1ms/sample - loss: 2.2642 - acc: 0.1495 - val_loss: 2.1794 - val_acc: 0.2021
Epoch 3/10
1438/1438 [==============================] - 4s 2ms/sample - loss: 2.2063 - acc: 0.1822 - val_loss: 2.1506 - val_acc: 0.1979
Epoch 4/10
1438/1438 [==============================] - 4s 2ms/sample - loss: 2.1890 - acc: 0.2017 - val_loss: 2.1144 - val_acc: 0.2229
Epoch 5/10
1438/1438 [==============================] - 3s 2ms/sample - loss: 2.1330 - acc: 0.2211 - val_loss: 2.1030 - val_acc: 0.1854
Epoch 6/10
1438/1438 [==============================] - 2s 2ms/sample - loss: 2.1050 - acc: 0.2427 - val_loss: 2.0784 - val_acc: 0.2000
Epoch 7/10
1438/1438 [==============================] - 2s 2ms/sample - loss: 2.0841 - acc: 0.2427 - val_loss: 2.0601 - val_acc: 0.2271
E

In [42]:
y_val_nn = np.argmax(y_val, axis=1) # Convert one-hot to index
y_pred = model.predict_classes(X_val)
print(classification_report(y_val_nn, y_pred))

              precision    recall  f1-score   support

           0       0.49      0.42      0.45        43
           1       0.26      0.40      0.31        43
           2       0.38      0.60      0.46        48
           3       0.43      0.11      0.17        56
           4       0.62      0.23      0.34        43
           5       0.48      0.23      0.31        48
           6       0.56      0.35      0.43        55
           7       0.58      0.12      0.21        56
           8       0.31      0.11      0.16        47
           9       0.15      0.66      0.24        41

    accuracy                           0.31       480
   macro avg       0.43      0.32      0.31       480
weighted avg       0.43      0.31      0.31       480



From what we can see normalising spectrograms is the way to go. Let's use it by default

### Best model

In [43]:
%%time
X_train, X_val, X_test, y_train, y_val, y_test = data_preparation.split_train_test_baseline_spectrograms(norm_spects, labels_digits)
X_train = np.concatenate([X_train, X_val], axis=0)
y_train = np.concatenate([y_train, y_val], axis=0)
clf1 = SVC(kernel='rbf', class_weight='balanced', gamma="auto")
clf1 = clf1.fit(X_train, y_train)
y_pred = clf1.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.95      0.88        44
           1       0.88      0.86      0.87        49
           2       0.76      0.91      0.83        56
           3       0.97      0.70      0.81        43
           4       0.98      0.90      0.94        49
           5       0.90      0.87      0.88        52
           6       0.69      0.83      0.75        42
           7       0.79      0.98      0.88        47
           8       0.97      0.72      0.83        50
           9       0.90      0.79      0.84        48

    accuracy                           0.85       480
   macro avg       0.87      0.85      0.85       480
weighted avg       0.87      0.85      0.85       480

CPU times: user 30.5 s, sys: 225 ms, total: 30.7 s
Wall time: 31 s


## Speakers
### SVD

In [44]:
X_train, X_val, X_test, y_train, y_val, y_test = data_preparation.split_train_test_baseline_spectrograms(norm_spects, labels_speakers)

In [45]:
%%time
clf1 = SVC(kernel='rbf', class_weight='balanced', gamma="auto")
clf1 = clf1.fit(X_train, y_train)

CPU times: user 10.1 s, sys: 101 ms, total: 10.2 s
Wall time: 10.4 s


In [46]:
%%time
y_pred = clf1.predict(X_val)
print(classification_report(y_val, y_pred))

              precision    recall  f1-score   support

         ale       0.91      0.95      0.93        21
      alinda       0.90      0.90      0.90        20
        gian       1.00      1.00      1.00        20
     jackson       1.00      1.00      1.00        86
      khaled       0.91      1.00      0.95        21
     nicolas       0.98      1.00      0.99       103
        theo       0.86      0.87      0.86       105
    yweweler       0.92      0.87      0.89       104

    accuracy                           0.94       480
   macro avg       0.93      0.95      0.94       480
weighted avg       0.94      0.94      0.94       480

CPU times: user 3.96 s, sys: 29.4 ms, total: 3.99 s
Wall time: 4.08 s


### CNN

For neural networks it is not possible to pass the labels as-is: we need to transform them in numbers. The safest way is through one-hot encoding

In [47]:
y, target_names = data_preparation.transform_categorical_y(labels_speakers)

In [48]:
X_train, X_val, X_test, y_train, y_val, y_test, input_shape = data_preparation.split_train_test_nn(norm_spects, y, number_mode=False)

In [49]:
model = cnn_models.paper_architecture(8, input_shape=input_shape)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_4 (Conv2D)            (None, 63, 27, 32)        544       
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 30, 12, 32)        0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 14, 5, 64)         32832     
_________________________________________________________________
max_pooling2d_5 (MaxPooling2 (None, 6, 1, 64)          0         
_________________________________________________________________
flatten_2 (Flatten)          (None, 384)               0         
_________________________________________________________________
dense_6 (Dense)              (None, 80)                30800     
_________________________________________________________________
dropout_2 (Dropout)          (None, 80)                0         
__________

In [50]:
%%time
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val, y_val))

Train on 1438 samples, validate on 480 samples
Epoch 1/10
1438/1438 [==============================] - 2s 2ms/sample - loss: 1.9019 - acc: 0.2364 - val_loss: 1.8318 - val_acc: 0.2625
Epoch 2/10
1438/1438 [==============================] - 2s 1ms/sample - loss: 1.8193 - acc: 0.2949 - val_loss: 1.7708 - val_acc: 0.2854
Epoch 3/10
1438/1438 [==============================] - 2s 1ms/sample - loss: 1.7627 - acc: 0.3519 - val_loss: 1.6811 - val_acc: 0.4396
Epoch 4/10
1438/1438 [==============================] - 2s 1ms/sample - loss: 1.6879 - acc: 0.4124 - val_loss: 1.5829 - val_acc: 0.5604
Epoch 5/10
1438/1438 [==============================] - 2s 1ms/sample - loss: 1.5869 - acc: 0.4513 - val_loss: 1.4760 - val_acc: 0.5813
Epoch 6/10
1438/1438 [==============================] - 2s 1ms/sample - loss: 1.5152 - acc: 0.4861 - val_loss: 1.3654 - val_acc: 0.5708
Epoch 7/10
1438/1438 [==============================] - 2s 1ms/sample - loss: 1.4035 - acc: 0.5216 - val_loss: 1.2527 - val_acc: 0.6292
E

In [51]:
Y_val_nn = np.argmax(y_val, axis=1) # Convert one-hot to index
y_pred = model.predict_classes(X_val)
print(classification_report(Y_val_nn, y_pred))

              precision    recall  f1-score   support

           0       0.55      0.57      0.56        21
           1       0.00      0.00      0.00        20
           2       0.50      0.15      0.23        20
           3       0.84      0.77      0.80        86
           4       1.00      0.10      0.17        21
           5       0.53      0.36      0.43       103
           6       0.59      0.79      0.68       105
           7       0.46      0.70      0.56       104

    accuracy                           0.57       480
   macro avg       0.56      0.43      0.43       480
weighted avg       0.58      0.57      0.55       480



#### Paper - batch_normalisation=True

In [52]:
model = cnn_models.paper_architecture(8, input_shape, batch_normalisation=True)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_6 (Conv2D)            (None, 63, 27, 32)        544       
_________________________________________________________________
batch_normalization_v1 (Batc (None, 63, 27, 32)        128       
_________________________________________________________________
max_pooling2d_6 (MaxPooling2 (None, 30, 12, 32)        0         
_________________________________________________________________
conv2d_7 (Conv2D)            (None, 14, 5, 64)         32832     
_________________________________________________________________
batch_normalization_v1_1 (Ba (None, 14, 5, 64)         256       
_________________________________________________________________
max_pooling2d_7 (MaxPooling2 (None, 6, 1, 64)          0         
_________________________________________________________________
flatten_3 (Flatten)          (None, 384)               0         
__________

In [53]:
%%time
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val, y_val))

Train on 1438 samples, validate on 480 samples
Epoch 1/10
1438/1438 [==============================] - 9s 6ms/sample - loss: 1.7556 - acc: 0.4242 - val_loss: 1.7411 - val_acc: 0.4354
Epoch 2/10
1438/1438 [==============================] - 7s 5ms/sample - loss: 1.0340 - acc: 0.6933 - val_loss: 1.6581 - val_acc: 0.3979
Epoch 3/10
1438/1438 [==============================] - 8s 5ms/sample - loss: 0.8612 - acc: 0.7288 - val_loss: 1.7820 - val_acc: 0.4000
Epoch 4/10
1438/1438 [==============================] - 9s 6ms/sample - loss: 0.7052 - acc: 0.7830 - val_loss: 1.9644 - val_acc: 0.3854
Epoch 5/10
1438/1438 [==============================] - 8s 5ms/sample - loss: 0.5870 - acc: 0.8220 - val_loss: 1.8893 - val_acc: 0.3938
CPU times: user 1min 15s, sys: 13.5 s, total: 1min 28s
Wall time: 41.6 s


In [54]:
y_pred = model.predict_classes(X_val)
print(classification_report(Y_val_nn, y_pred))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00        21
           1       0.00      0.00      0.00        20
           2       0.00      0.00      0.00        20
           3       0.46      0.99      0.63        86
           4       0.00      0.00      0.00        21
           5       0.71      0.10      0.17       103
           6       0.34      0.91      0.49       105
           7       0.00      0.00      0.00       104

    accuracy                           0.40       480
   macro avg       0.19      0.25      0.16       480
weighted avg       0.31      0.40      0.26       480



/Users/kappa/opt/miniconda3/envs/dsim/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### Best model

In [55]:
%%time
X_train, X_val, X_test, y_train, y_val, y_test = data_preparation.split_train_test_baseline_spectrograms(norm_spects, labels_speakers)
X_train = np.concatenate([X_train, X_val], axis=0)
y_train = np.concatenate([y_train, y_val], axis=0)
clf1 = SVC(kernel='rbf', class_weight='balanced', gamma="auto")
clf1 = clf1.fit(X_train, y_train)
y_pred = clf1.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

         ale       1.00      0.96      0.98        24
      alinda       0.90      1.00      0.95        19
        gian       0.96      1.00      0.98        24
     jackson       1.00      1.00      1.00       121
      khaled       0.93      1.00      0.97        14
     nicolas       1.00      1.00      1.00        89
        theo       0.95      0.86      0.90        98
    yweweler       0.89      0.96      0.92        91

    accuracy                           0.96       480
   macro avg       0.96      0.97      0.96       480
weighted avg       0.96      0.96      0.96       480

CPU times: user 21.7 s, sys: 332 ms, total: 22 s
Wall time: 22.7 s


# Data augmentation
## Speaker

In [56]:
nsamples, nx, ny = X_train_speaker.shape
X_train_speaker_2d = X_train_speaker.reshape((nsamples, nx * ny))

In [57]:
nsamples, nx, ny = X_val_speaker.shape
X_val_speaker_2d = X_val_speaker.reshape((nsamples, nx * ny))

In [58]:
%%time
# Switch to LinearSVC because SVC with RBF kernel takes a lot of time
from sklearn.svm import LinearSVC
clf1 = LinearSVC(class_weight='balanced')
clf1 = clf1.fit(X_train_speaker_2d, y_train_speaker)

CPU times: user 1min 53s, sys: 2.23 s, total: 1min 56s
Wall time: 2min 1s


/Users/kappa/opt/miniconda3/envs/dsim/lib/python3.7/site-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [59]:
y_pred = clf1.predict(X_val_speaker_2d)
print(classification_report(y_val_speaker, y_pred))

              precision    recall  f1-score   support

         ale       0.93      0.86      0.89        86
      alinda       1.00      0.92      0.96        85
        gian       0.95      0.94      0.94        82
     jackson       0.98      0.98      0.98       596
      khaled       0.93      0.93      0.93        83
     nicolas       0.96      0.97      0.96       545
        theo       0.79      0.79      0.79       579
    yweweler       0.78      0.79      0.78       534

    accuracy                           0.89      2590
   macro avg       0.91      0.90      0.91      2590
weighted avg       0.89      0.89      0.89      2590



### CNNs

In [60]:
from sklearn.preprocessing import OneHotEncoder
enc = OneHotEncoder()
y_train_speaker = enc.fit_transform(y_train_speaker.reshape(-1, 1)).toarray()
y_val_speaker = enc.transform(y_val_speaker.reshape(-1, 1)).toarray()
y_test_speaker = enc.transform(y_test_speaker.reshape(-1, 1)).toarray()
label_0 = enc.inverse_transform(np.array([1, 0, 0, 0, 0, 0, 0, 0]).reshape(1, -1))[0][0]
label_1 = enc.inverse_transform(np.array([0, 1, 0, 0, 0, 0, 0, 0]).reshape(1, -1))[0][0]
label_2 = enc.inverse_transform(np.array([0, 0, 1, 0, 0, 0, 0, 0]).reshape(1, -1))[0][0]
label_3 = enc.inverse_transform(np.array([0, 0, 0, 1, 0, 0, 0, 0]).reshape(1, -1))[0][0]
label_4 = enc.inverse_transform(np.array([0, 0, 0, 0, 1, 0, 0, 0]).reshape(1, -1))[0][0]
label_5 = enc.inverse_transform(np.array([0, 0, 0, 0, 0, 1, 0, 0]).reshape(1, -1))[0][0]
label_6 = enc.inverse_transform(np.array([0, 0, 0, 0, 0, 0, 1, 0]).reshape(1, -1))[0][0]
label_7 = enc.inverse_transform(np.array([0, 0, 0, 0, 0, 0, 0, 1]).reshape(1, -1))[0][0]
target_names = [label_0, label_1, label_2, label_3, label_4, label_5, label_6, label_7]

In [61]:
X_train_speaker = X_train_speaker.reshape(X_train_speaker.shape[0],
                                          X_train_speaker.shape[1],
                                          X_train_speaker.shape[2],
                                          1)
X_val_speaker = X_val_speaker.reshape(X_val_speaker.shape[0],
                                      X_val_speaker.shape[1],
                                      X_val_speaker.shape[2],
                                      1)
X_test_speaker = X_test_speaker.reshape(X_test_speaker.shape[0],
                                        X_test_speaker.shape[1],
                                        X_test_speaker.shape[2],
                                        1)

In [62]:
input_shape = (X_train_speaker.shape[1], X_train_speaker.shape[2], 1)

In [63]:
model = cnn_models.paper_architecture(8, input_shape=input_shape)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_8 (Conv2D)            (None, 63, 27, 32)        544       
_________________________________________________________________
max_pooling2d_8 (MaxPooling2 (None, 30, 12, 32)        0         
_________________________________________________________________
conv2d_9 (Conv2D)            (None, 14, 5, 64)         32832     
_________________________________________________________________
max_pooling2d_9 (MaxPooling2 (None, 6, 1, 64)          0         
_________________________________________________________________
flatten_4 (Flatten)          (None, 384)               0         
_________________________________________________________________
dense_12 (Dense)             (None, 80)                30800     
_________________________________________________________________
dropout_4 (Dropout)          (None, 80)                0         
__________

In [64]:
%%time
model.fit(X_train_speaker, y_train_speaker,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val_speaker, y_val_speaker))

Train on 10358 samples, validate on 2590 samples
Epoch 1/10
10358/10358 [==============================] - 23s 2ms/sample - loss: 1.7699 - acc: 0.3083 - val_loss: 1.5724 - val_acc: 0.5054
Epoch 2/10
10358/10358 [==============================] - 24s 2ms/sample - loss: 1.4804 - acc: 0.4800 - val_loss: 1.2220 - val_acc: 0.5525
Epoch 3/10
10358/10358 [==============================] - 23s 2ms/sample - loss: 1.2099 - acc: 0.5571 - val_loss: 0.9510 - val_acc: 0.6734
Epoch 4/10
10358/10358 [==============================] - 34s 3ms/sample - loss: 1.0140 - acc: 0.6167 - val_loss: 0.8133 - val_acc: 0.6907
Epoch 5/10
10358/10358 [==============================] - 22s 2ms/sample - loss: 0.8770 - acc: 0.6640 - val_loss: 0.8376 - val_acc: 0.6421
Epoch 6/10
10358/10358 [==============================] - 19s 2ms/sample - loss: 0.7540 - acc: 0.7086 - val_loss: 0.5789 - val_acc: 0.7757
Epoch 7/10
10358/10358 [==============================] - 33s 3ms/sample - loss: 0.6494 - acc: 0.7420 - val_loss: 0.5

In [65]:
Y_val_nn = np.argmax(y_val_speaker, axis=1) # Convert one-hot to index
y_pred = model.predict_classes(X_val_speaker)
print(classification_report(Y_val_nn, y_pred, target_names=target_names))

              precision    recall  f1-score   support

         ale       0.81      0.78      0.79        86
      alinda       0.69      0.45      0.54        85
        gian       0.58      0.70      0.63        82
     jackson       0.81      1.00      0.89       596
      khaled       0.80      0.53      0.64        83
     nicolas       0.72      0.85      0.78       545
        theo       0.95      0.50      0.65       579
    yweweler       0.62      0.71      0.66       534

    accuracy                           0.75      2590
   macro avg       0.75      0.69      0.70      2590
weighted avg       0.77      0.75      0.74      2590



### Batch_normalization = True

In [66]:
model = cnn_models.paper_architecture(8, input_shape=input_shape, batch_normalisation=True)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_10 (Conv2D)           (None, 63, 27, 32)        544       
_________________________________________________________________
batch_normalization_v1_4 (Ba (None, 63, 27, 32)        128       
_________________________________________________________________
max_pooling2d_10 (MaxPooling (None, 30, 12, 32)        0         
_________________________________________________________________
conv2d_11 (Conv2D)           (None, 14, 5, 64)         32832     
_________________________________________________________________
batch_normalization_v1_5 (Ba (None, 14, 5, 64)         256       
_________________________________________________________________
max_pooling2d_11 (MaxPooling (None, 6, 1, 64)          0         
_________________________________________________________________
flatten_5 (Flatten)          (None, 384)               0         
__________

In [67]:
%%time
model.fit(X_train_speaker, y_train_speaker,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val_speaker, y_val_speaker))

Train on 10358 samples, validate on 2590 samples
Epoch 1/10
10358/10358 [==============================] - 72s 7ms/sample - loss: 1.1763 - acc: 0.5926 - val_loss: 0.8747 - val_acc: 0.7185
Epoch 2/10
10358/10358 [==============================] - 57s 6ms/sample - loss: 0.7152 - acc: 0.7333 - val_loss: 0.8116 - val_acc: 0.7004
Epoch 3/10
10358/10358 [==============================] - 61s 6ms/sample - loss: 0.5687 - acc: 0.7837 - val_loss: 0.3933 - val_acc: 0.8571
Epoch 4/10
10358/10358 [==============================] - 74s 7ms/sample - loss: 0.4963 - acc: 0.8119 - val_loss: 0.4268 - val_acc: 0.8255
Epoch 5/10
10358/10358 [==============================] - 60s 6ms/sample - loss: 0.4388 - acc: 0.8314 - val_loss: 0.8710 - val_acc: 0.6768
Epoch 6/10
10358/10358 [==============================] - 66s 6ms/sample - loss: 0.4063 - acc: 0.8457 - val_loss: 0.3571 - val_acc: 0.8591
Epoch 7/10
10358/10358 [==============================] - 56s 5ms/sample - loss: 0.3688 - acc: 0.8629 - val_loss: 0.3

In [68]:
y_pred = model.predict_classes(X_val_speaker)
print(classification_report(Y_val_nn, y_pred, target_names=target_names))

              precision    recall  f1-score   support

         ale       0.90      0.98      0.94        86
      alinda       0.91      0.93      0.92        85
        gian       0.93      0.87      0.90        82
     jackson       0.94      1.00      0.97       596
      khaled       0.93      0.93      0.93        83
     nicolas       1.00      0.92      0.96       545
        theo       0.78      0.95      0.86       579
    yweweler       0.94      0.72      0.82       534

    accuracy                           0.91      2590
   macro avg       0.92      0.91      0.91      2590
weighted avg       0.91      0.91      0.90      2590



### Different architecture
Let's change a bit the architecture and see if we can improve scores:

In [69]:
model = cnn_models.custom_cnn(8, input_shape=input_shape)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_12 (Conv2D)           (None, 63, 27, 32)        544       
_________________________________________________________________
conv2d_13 (Conv2D)           (None, 30, 12, 64)        32832     
_________________________________________________________________
max_pooling2d_12 (MaxPooling (None, 14, 5, 64)         0         
_________________________________________________________________
flatten_6 (Flatten)          (None, 4480)              0         
_________________________________________________________________
dense_18 (Dense)             (None, 128)               573568    
_________________________________________________________________
dropout_6 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_19 (Dense)             (None, 8)                 1032      
Total para

In [70]:
%%time
model.fit(X_train_speaker, y_train_speaker,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val_speaker, y_val_speaker))

Train on 10358 samples, validate on 2590 samples
Epoch 1/10
10358/10358 [==============================] - 40s 4ms/sample - loss: 1.4051 - acc: 0.4826 - val_loss: 1.1135 - val_acc: 0.6019
Epoch 2/10
10358/10358 [==============================] - 74s 7ms/sample - loss: 0.8388 - acc: 0.6935 - val_loss: 0.6079 - val_acc: 0.7722
Epoch 3/10
10358/10358 [==============================] - 91s 9ms/sample - loss: 0.5335 - acc: 0.7962 - val_loss: 0.3671 - val_acc: 0.8598
Epoch 4/10
10358/10358 [==============================] - 35s 3ms/sample - loss: 0.3754 - acc: 0.8531 - val_loss: 0.2500 - val_acc: 0.9008
Epoch 5/10
10358/10358 [==============================] - 37s 4ms/sample - loss: 0.2962 - acc: 0.8845 - val_loss: 0.2317 - val_acc: 0.9050
Epoch 6/10
10358/10358 [==============================] - 36s 4ms/sample - loss: 0.2471 - acc: 0.9003 - val_loss: 0.1753 - val_acc: 0.9344
Epoch 7/10
10358/10358 [==============================] - 36s 4ms/sample - loss: 0.2166 - acc: 0.9144 - val_loss: 0.1

In [71]:
y_pred = model.predict_classes(X_val_speaker)
print(classification_report(Y_val_nn, y_pred, target_names=target_names))

              precision    recall  f1-score   support

         ale       1.00      0.97      0.98        86
      alinda       1.00      0.98      0.99        85
        gian       0.95      0.95      0.95        82
     jackson       1.00      0.99      1.00       596
      khaled       0.98      0.98      0.98        83
     nicolas       0.99      0.99      0.99       545
        theo       0.85      0.99      0.91       579
    yweweler       0.97      0.81      0.89       534

    accuracy                           0.95      2590
   macro avg       0.97      0.96      0.96      2590
weighted avg       0.96      0.95      0.95      2590



### Best model
Based on the f1-score, the best model is the "custom cnn" one. Let's see its result on the test set:

In [72]:
%%time
X_train = np.concatenate([X_train_speaker, X_val_speaker], axis=0)
y_train = np.concatenate([y_train_speaker, y_val_speaker], axis=0)
model = cnn_models.custom_cnn(8, input_shape=input_shape)
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_14 (Conv2D)           (None, 63, 27, 32)        544       
_________________________________________________________________
conv2d_15 (Conv2D)           (None, 30, 12, 64)        32832     
_________________________________________________________________
max_pooling2d_13 (MaxPooling (None, 14, 5, 64)         0         
_________________________________________________________________
flatten_7 (Flatten)          (None, 4480)              0         
_________________________________________________________________
dense_20 (Dense)             (None, 128)               573568    
_________________________________________________________________
dropout_7 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_21 (Dense)             (None, 8)                 1032      
Total para

In [73]:
y_test_nn = np.argmax(y_test_speaker, axis=1)
y_pred = model.predict_classes(X_test_speaker)
print(classification_report(y_test_nn, y_pred, target_names=target_names))

              precision    recall  f1-score   support

         ale       1.00      0.90      0.95        30
      alinda       1.00      1.00      1.00        30
        gian       0.97      1.00      0.98        30
     jackson       0.91      1.00      0.95        30
      khaled       1.00      0.87      0.93        30
     nicolas       1.00      1.00      1.00        30
        theo       0.88      1.00      0.94        30
    yweweler       1.00      0.97      0.98        30

    accuracy                           0.97       240
   macro avg       0.97      0.97      0.97       240
weighted avg       0.97      0.97      0.97       240



In [75]:
model.save("../best_models/speaker_recognition.h5")

## Digits

In [76]:
nsamples, nx, ny = X_train_digit.shape
X_train_digit_2d = X_train_digit.reshape((nsamples, nx * ny))
nsamples, nx, ny = X_val_digit.shape
X_val_digit_2d = X_val_digit.reshape((nsamples, nx * ny))

In [77]:
%%time
# Switch to LinearSVC because SVC with RBF kernel takes a lot of time
clf1 = LinearSVC(class_weight='balanced')
clf1 = clf1.fit(X_train_digit_2d, y_train_digit)
y_pred = clf1.predict(X_val_digit_2d)
print(classification_report(y_val_digit, y_pred))

/Users/kappa/opt/miniconda3/envs/dsim/lib/python3.7/site-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


              precision    recall  f1-score   support

           0       0.78      0.74      0.75       438
           1       0.57      0.73      0.64       456
           2       0.67      0.62      0.65       455
           3       0.60      0.52      0.56       428
           4       0.79      0.65      0.72       485
           5       0.73      0.73      0.73       464
           6       0.63      0.62      0.62       442
           7       0.52      0.64      0.58       490
           8       0.70      0.73      0.71       499
           9       0.76      0.66      0.71       459

    accuracy                           0.67      4616
   macro avg       0.68      0.66      0.67      4616
weighted avg       0.68      0.67      0.67      4616

CPU times: user 10min 31s, sys: 12.6 s, total: 10min 43s
Wall time: 13min 31s


### CNNs

In [78]:
X_train_digit = X_train_digit.reshape(X_train_digit.shape[0], X_train_digit.shape[1], X_train_digit.shape[2], 1)
X_val_digit = X_val_digit.reshape(X_val_digit.shape[0], X_val_digit.shape[1], X_val_digit.shape[2], 1)
X_test_digit = X_test_digit.reshape(X_test_digit.shape[0], X_test_digit.shape[1], X_test_digit.shape[2], 1)
y_train_digit = tf.keras.utils.to_categorical(y_train_digit, 10)
y_test_digit = tf.keras.utils.to_categorical(y_test_digit, 10)
y_val_digit = tf.keras.utils.to_categorical(y_val_digit, 10)

In [79]:
input_shape = (X_train_digit.shape[1], X_train_digit.shape[2], 1)
input_shape

(128, 57, 1)

#### Paper

In [80]:
model = cnn_models.paper_architecture(10, input_shape=input_shape, batch_normalisation=True)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_16 (Conv2D)           (None, 63, 27, 32)        544       
_________________________________________________________________
batch_normalization_v1_8 (Ba (None, 63, 27, 32)        128       
_________________________________________________________________
max_pooling2d_14 (MaxPooling (None, 30, 12, 32)        0         
_________________________________________________________________
conv2d_17 (Conv2D)           (None, 14, 5, 64)         32832     
_________________________________________________________________
batch_normalization_v1_9 (Ba (None, 14, 5, 64)         256       
_________________________________________________________________
max_pooling2d_15 (MaxPooling (None, 6, 1, 64)          0         
_________________________________________________________________
flatten_8 (Flatten)          (None, 384)               0         
__________

In [81]:
%%time
model.fit(X_train_digit, y_train_digit,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val_digit, y_val_digit))

Train on 18462 samples, validate on 4616 samples
Epoch 1/10
18462/18462 [==============================] - 96s 5ms/sample - loss: 1.6812 - acc: 0.4126 - val_loss: 1.1362 - val_acc: 0.6237
Epoch 2/10
18462/18462 [==============================] - 89s 5ms/sample - loss: 1.1379 - acc: 0.6070 - val_loss: 0.8891 - val_acc: 0.7075
Epoch 3/10
18462/18462 [==============================] - 87s 5ms/sample - loss: 0.9713 - acc: 0.6669 - val_loss: 0.7421 - val_acc: 0.7517
Epoch 4/10
18462/18462 [==============================] - 166s 9ms/sample - loss: 0.8552 - acc: 0.7072 - val_loss: 0.7762 - val_acc: 0.7416
Epoch 5/10
18462/18462 [==============================] - 348s 19ms/sample - loss: 0.7833 - acc: 0.7358 - val_loss: 0.6693 - val_acc: 0.7734
Epoch 6/10
18462/18462 [==============================] - 209s 11ms/sample - loss: 0.7318 - acc: 0.7509 - val_loss: 0.5528 - val_acc: 0.8226
Epoch 7/10
18462/18462 [==============================] - 172s 9ms/sample - loss: 0.6905 - acc: 0.7656 - val_los

In [82]:
Y_val = np.argmax(y_val_digit, axis=1) # Convert one-hot to index
y_pred = model.predict_classes(X_val_digit)
print(classification_report(Y_val, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.94      0.88       438
           1       0.59      0.93      0.72       456
           2       0.98      0.57      0.72       455
           3       0.68      0.84      0.75       428
           4       0.90      0.80      0.84       485
           5       0.86      0.74      0.80       464
           6       0.90      0.81      0.85       442
           7       0.97      0.72      0.82       490
           8       0.92      0.83      0.87       499
           9       0.73      0.90      0.80       459

    accuracy                           0.81      4616
   macro avg       0.83      0.81      0.81      4616
weighted avg       0.84      0.81      0.81      4616



#### Custom

In [83]:
model = cnn_models.custom_cnn(10, input_shape=input_shape)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_18 (Conv2D)           (None, 63, 27, 32)        544       
_________________________________________________________________
conv2d_19 (Conv2D)           (None, 30, 12, 64)        32832     
_________________________________________________________________
max_pooling2d_16 (MaxPooling (None, 14, 5, 64)         0         
_________________________________________________________________
flatten_9 (Flatten)          (None, 4480)              0         
_________________________________________________________________
dense_25 (Dense)             (None, 128)               573568    
_________________________________________________________________
dropout_9 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_26 (Dense)             (None, 10)                1290      
Total para

In [84]:
%%time
model.fit(X_train_digit, y_train_digit,
          batch_size=32,
          epochs=10,
          verbose=1,
          callbacks=[callback],
          validation_data=(X_val_digit, y_val_digit))

Train on 18462 samples, validate on 4616 samples
Epoch 1/10
18462/18462 [==============================] - 67s 4ms/sample - loss: 1.5261 - acc: 0.4572 - val_loss: 1.1146 - val_acc: 0.6098
Epoch 2/10
18462/18462 [==============================] - 65s 4ms/sample - loss: 1.0347 - acc: 0.6393 - val_loss: 0.8179 - val_acc: 0.7179
Epoch 3/10
18462/18462 [==============================] - 66s 4ms/sample - loss: 0.8470 - acc: 0.6988 - val_loss: 0.7932 - val_acc: 0.7240
Epoch 4/10
18462/18462 [==============================] - 66s 4ms/sample - loss: 0.7316 - acc: 0.7415 - val_loss: 0.6625 - val_acc: 0.7727
Epoch 5/10
18462/18462 [==============================] - 65s 4ms/sample - loss: 0.6629 - acc: 0.7638 - val_loss: 0.6325 - val_acc: 0.7886
Epoch 6/10
18462/18462 [==============================] - 66s 4ms/sample - loss: 0.6123 - acc: 0.7825 - val_loss: 0.5447 - val_acc: 0.8078
Epoch 7/10
18462/18462 [==============================] - 66s 4ms/sample - loss: 0.5566 - acc: 0.8056 - val_loss: 0.4

In [85]:
y_pred = model.predict_classes(X_val_digit)
print(classification_report(Y_val, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.94      0.88       438
           1       0.75      0.87      0.81       456
           2       0.81      0.86      0.83       455
           3       0.82      0.71      0.76       428
           4       0.94      0.87      0.90       485
           5       0.88      0.86      0.87       464
           6       0.92      0.81      0.87       442
           7       0.94      0.85      0.89       490
           8       0.90      0.88      0.89       499
           9       0.80      0.89      0.84       459

    accuracy                           0.85      4616
   macro avg       0.86      0.85      0.85      4616
weighted avg       0.86      0.85      0.86      4616



### Best model
Based on F1-Score the best model is once again the custom paper architecture:

In [86]:
%%time
X_train = np.concatenate([X_train_digit, X_val_digit], axis=0)
y_train = np.concatenate([y_train_digit, y_val_digit], axis=0)
model = cnn_models.custom_cnn(10, input_shape=input_shape)
model.fit(X_train, y_train,
          batch_size=32,
          epochs=10,
          verbose=1)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_20 (Conv2D)           (None, 63, 27, 32)        544       
_________________________________________________________________
conv2d_21 (Conv2D)           (None, 30, 12, 64)        32832     
_________________________________________________________________
max_pooling2d_17 (MaxPooling (None, 14, 5, 64)         0         
_________________________________________________________________
flatten_10 (Flatten)         (None, 4480)              0         
_________________________________________________________________
dense_27 (Dense)             (None, 128)               573568    
_________________________________________________________________
dropout_10 (Dropout)         (None, 128)               0         
_________________________________________________________________
dense_28 (Dense)             (None, 10)                1290      
Total para

In [87]:
y_test_nn = np.argmax(y_test_digit, axis=1)
y_pred = model.predict_classes(X_test_digit)
print(classification_report(y_test_nn, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.93      0.97        30
           1       0.97      0.93      0.95        30
           2       0.97      1.00      0.98        30
           3       0.97      0.93      0.95        30
           4       0.97      1.00      0.98        30
           5       0.78      0.93      0.85        30
           6       0.94      1.00      0.97        30
           7       0.90      0.93      0.92        30
           8       0.97      0.93      0.95        30
           9       0.96      0.77      0.85        30

    accuracy                           0.94       300
   macro avg       0.94      0.94      0.94       300
weighted avg       0.94      0.94      0.94       300



In [88]:
model.save("../best_models/digit_recognition.h5")

# Test model 

In [ ]:
import sounddevice as sd
import subprocess

import time
import librosa

import IPython.display as ipd

import os
from scipy.io import wavfile as wav

In [ ]:
def pad_zeros_single_rec(rec, max_y):
    rec = np.array(rec)
    diff_in_y = max_y - rec.shape[0]
    if diff_in_y > 0:
        half_diff = int(diff_in_y/2)
        remaining_diff = diff_in_y-half_diff
        v = np.pad(rec, (half_diff, remaining_diff), 'constant', constant_values=0)
        return v
    else:
        return rec

In [ ]:
def create_recording(duration, rec_rate, name = "test.wav", output_dir = "test/"):
    print("Ready in 3...", end = "")
    time.sleep(1)
    print("2...", end = "")
    time.sleep(1)
    print("1...")
    time.sleep(1)
    print("Go.")
    rec = sd.rec(int(duration * rec_rate), samplerate=rec_rate, channels=1, blocking=True)
    print("Playing the recording.")
    sd.play(rec, rec_rate)

    # after hearing the recording, decide whether to record it again or continue to next number
    # if you type anything, record again
    # if you press enter, save current recording & go to next number
    ok = input("OK?")
    if ok == "":
        librosa.output.write_wav(output_dir+name, rec, rec_rate)
        return rec
    ipd.clear_output(wait=True)
    create_recording(duration, rec_rate)

In [ ]:
def trim_audio(file, input_dir="test/", output_dir="test/", db=-48):

    if not os.path.isdir(input_dir):
        print(f"There should be an input \"{input_dir}\" directory.")
        sys.exit(0)
    
    # create output directory if not there yet
    if not os.path.isdir(output_dir):
        os.makedirs(output_dir)
        
    temp1 = output_dir+"temp1.wav"
    temp2 = output_dir+"temp2.wav"
    temp3 = output_dir+"temp3.wav"
 
    subprocess.run(["ffmpeg", "-y", "-i", input_dir+file, "-af", f"silenceremove=1:0:{db}dB", temp1])
    subprocess.run(["ffmpeg", "-y", "-i", temp1, "-af", "areverse", temp2])
    subprocess.run(["ffmpeg", "-y", "-i", temp2, "-af", f"silenceremove=1:0.1:{db}dB", temp3])
    subprocess.run(["ffmpeg", "-y", "-i", temp3, "-af", "areverse", output_dir+file])
    
    os.remove(temp1)
    os.remove(temp2)
    os.remove(temp3)

In [ ]:
def test_NN(nn, max_y, target_names, answer = None, duration=2, rec_rate=8000, directory = "test/", filename = "test.wav"):
    create_recording(duration, rec_rate, filename, directory)   
    ipd.clear_output()
    trim_audio(filename, directory, directory)
    # _, rec = wav.read(directory + "/" + filename)
    rec, _ = librosa.core.load(directory + "/" + filename, sr = rec_rate)
    rec = pad_zeros_single_rec(rec, max_y)
    # sd.play(rec, rec_rate)
    rec = data_preparation.compute_spectrogram(rec, normalize=True)
    rec = rec[np.newaxis,:,:,np.newaxis]
    preds = nn.predict_classes(rec)
    print("Model prediction: {}".format(target_names[preds[0]]))
    if answer is not None:
        print(f"Correct answer {answer}")
    return preds

In [ ]:
max_y = len(data_augm_pad_recordings[1])

In [ ]:
pred = test_NN(model, max_y, target_names, answer = "gian")

# TO DO:
- [x] Set random seed
- [x] Use only original recordings in test set of augmented scenario
- [x] Use proper validation set for picking best models and params
- [x] Data augmentation also for digit recognition
- [ ] Evaluate each best model on test set, after training it on x_train + x_test